tools/awslint/GUIDELINES.md

```diff
@@ -0,0 +1,425 @@
+
+## Resource Interface
+
+> <a id="resource-interface">awslint: resource-interface</a>
+
+Every AWS resource should have a resource interface `IFoo`.
+
+This interface represents both resources defined within the same stack (aka
+"internal" or "owned") or resources that are defined in different stack/app (aka
+"imported", "existing", "external" or "unowned"). Throughout this document we
+shall refer to these two types of resources as **"internal"** and
+**"external"**.
+
+> <a id="resource-interface-extends-construct">awslint: resource-interface-extends-construct</a>
+
+Resource interfaces should extend `cdk.IConstruct` in order to allow consumers
+to take advantage of construct capabilities such as unique IDs, paths, scopes, etc.
+
+> TODO: <a id="resource-ref-interface">awslint: resource-ref-interface</a>
+
+When resources are referenced anywhere in the API (e.g. in properties or methods
+of other resources or higher level constructs), the resource interface (`IFoo`)
+should be preferred over the concrete resource class (`Foo`). This will allow
+users to supply either internal or external resources.
+
+## Resource Attributes
+
+Every AWS resource has a set of "physical" runtime attributes such as ARN,
+physical names, URLs, etc. These attributes are commonly late-bound, which means
+they can only be resolved when AWS CloudFormation actually provisions the
+resource.
+
+Resource attributes almost always represent **string** values (URL, ARN, name).
+Sometimes they might also represent a **list of strings**.
+
+> <a id="resource-attribute">awslint: resource-attribute</a><br/>
+> <a id="resource-attribute-immutable">awslint: resource-attribute-immutable</a>
+
+All resource attributes must be represented as readonly properties of the
+resource interface. The names of the attributes must correspond to the
+CloudFormation resource attribute name.
+
+> TODO: <a id="resource-attribute-type">awslint: resource-attribute-type</a>
+
+Since attribute values can either be late-bound ("a promise to a string") or
+concrete ("a string"), the AWS CDK has a mechanism called "tokens" which allows
+codifying late-bound values into strings or string arrays. This approach was
+chosen in order to dramatically simplify the type-system and ergonomics of CDK
+code. As long as users treat these attributes as _opaque values_ (e.g. not try
+to parse them or manipulate them), they can be used interchangeably.
+
+For example, users can concatenate strings with tokens idiomatically:
+
+```ts
+`This is my bucket name: ${bucket.bucketName} and bucket ARN: ${bucket.bucketArn}`
+```
+
+Even though `bucketName` and `bucketArn` will only be resolved during
+deployment, the CDK will identify those as tokens and will convert this string
+into an `{ "Fn::Join" }` expression which includes the relevant intrinsic
+functions.
+
+Resource attributes should use a type that corresponds to the __resolved__ AWS
+CloudFormation type (e.g. `string`, `string[]`).
```

Comment:
For strings it is not needed since the generated attributes are [already strings](https://github.com/awslabs/aws-cdk/blob/master/tools/cfn2ts/lib/codegen.ts#L291), but that's not the case (yet) for string arrays or (in the future) numbers (#1455).